paging/common/src/main/kotlin/androidx/paging/InvalidatingPagingSourceFactory.kt

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.paging
+
+import androidx.annotation.VisibleForTesting
+
+/**
+ * A PagingSource Factory that can be kept as a reference and stores a list of paging sources
+ * created with the factory. Includes invalidate() function to invalidate the list of [PagingSource]
+ * when data changes.
+ */
+abstract class InvalidatingPagingSourceFactory<Key: Any, Value: Any> : () ->  PagingSource<Key, Value>{
+
+    @VisibleForTesting
+    internal val listOfPagingSources = mutableListOf<PagingSource<Key, Value>>();
+
+    /**
+     * Returns a PagingSource
+     *
+     * Implement the logic to create a PagingSource
+     */
+    abstract fun create() : PagingSource<Key, Value>;
```

Comment:
This should at least be `internal` to prevent confusion for those using it incorrectly like in

```
Pager(...) {
    MyInvalidatingPagingSourceFactory.create()
}
```
but most likely even if it's internal people will still be able to do the incorrect thing since their usage would likely be in the same package.

A better approach might be to take in the original factory using a property in constructor like so:

abstract class InvalidatingPagingSourceFactory<Key: Any, Value: Any>(
    val pagingSourceFactory: () -> PagingSource<Key, Value>
) {
    ...
    final override fun invoke(): ... {
        return pagingSourceFactory().also { ... }
    }
}
```